### PR TITLE
Fix bug with reconnection failing during high request load

### PR DESCRIPTION
**Pull request by krka**

### CODE REVIEW

File: folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java [hunk 1]
```diff
@@ -152,12 +152,17 @@ protected void initChannel(final Channel ch) throws Exception {
     connectFuture.addListener(
         (ChannelFutureListener)
             future -> {
+              final Channel channel = future.channel();
```
Comment (krka) [inline folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java:155]: not strictly needed for this change, but probably a good extra safety measure

File: folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java [hunk 2]
```diff
@@ -166,12 +171,12 @@ protected void initChannel(final Channel ch) throws Exception {
...
                 clientFuture.complete(client);
               } else {
-                future.channel().close();
+                channel.close();
                 clientFuture.completeExceptionally(future.cause());
               }
             });
 
-    return Utils.onExecutor(clientFuture, executor);
+    return clientFuture;
```
Comment (krka) [inline folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java:179]: this is the main bugfix - don't use the reply-executor for connection callbacks

File: folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java [hunk 3]
```diff
@@ -158,12 +158,10 @@ private void retry() {
       future.whenComplete(
           (newClient, t) -> {
             if (t != null) {
-              if (t instanceof CompletionException) {
-                if (t.getCause() instanceof MemcacheAuthenticationException) {
-                  connectionFailure = t.getCause();
-                  shutdown();
-                  return;
-                }
+              if (t instanceof CompletionException && t.getCause() instanceof MemcacheAuthenticationException) {
```
Comment (krka) [inline folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java]: no logic change here

File: folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java [hunk 4]
```diff
@@ -198,25 +196,27 @@ private void retry() {
   }
 
   private void onFailure() {
+    if (!stayConnected) {
```
Comment (krka) [inline folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java:200]: just simplifying the code a bit

**Comment by alorlea** [comment]
Seems good :ok_hand:

**Comment by protocol7** [comment]
+1

**Pull request merged**